go/consensus/tendermint/full/full.go

```diff
@@ -1289,6 +1303,38 @@ func (t *fullService) lazyInit() error {
 		t.client = tmcli.New(t.node)
 		t.failMonitor = newFailMonitor(t.ctx, t.Logger, t.node.ConsensusState().Wait)
 
+		// Register a halt hook that handles upgrades gracefully.
+		t.RegisterHaltHook(func(ctx context.Context, blockHeight int64, epoch beaconAPI.EpochTime, err error) {
+			if !errors.Is(err, upgradeAPI.ErrStopForUpgrade) {
+				return
+			}
+
+			// Mark this as a clean shutdown and request the node to stop gracefully.
+			t.failMonitor.markCleanShutdown()
+
+			// Wait before stopping to give time for P2P messages to propagate. Sleep for at least
+			// minUpgradeStopWaitPeriod or the configured commit timeout.
+			t.Logger.Info("waiting a bit before stopping the node for upgrade")
+			waitPeriod := minUpgradeStopWaitPeriod
+			if t.genesis.Consensus.Parameters.TimeoutCommit > waitPeriod {
+				waitPeriod = t.genesis.Consensus.Parameters.TimeoutCommit
+			}
+			time.Sleep(waitPeriod)
+
+			go func() {
+				// Sleep another period so there is some time between when consensus shuts down and
+				// when all the other services start shutting down.
+				time.Sleep(waitPeriod)
```

Comment:
Would it make more sense to use a different (static, not dependent on `TimeoutCommit`) period here?